### ISSUE TITLE
Transactions are no longer rolled back independently since 0.3.18

### ISSUE THREAD

**Issue by brunnerh** [typeorm/typeorm#11143]
### Issue description

If two transactions run at the same time and one is rolled back due to error, the result suggests that now both are being rolled back.

### Expected Behavior

Only the rolled back transaction should be affected (unless transactions interact in ways which require that they are summarily rolled back).

### Actual Behavior

Two transactions insert an entity into the same table.
One is rolled back.
Then the other is committed.

There are no entities in the table.

### Steps to reproduce

The core test logic is this:
```ts
const { promise: wait, resolve: endWait } = Promise.withResolvers<void>();

const t1 = transaction(dataSource, async em => {
	await em.save(new Entity1('a1'));
	endWait();
	await t2.catch(() => {});
});
const t2 = transaction(dataSource, async em => {
	await em.save(new Entity1('a2'));
	await wait;
	throw new Error('rollback transaction 2');
});

const [r1, r2] = await Promise.allSettled([t1, t2]);
const entities = await dataSource.getRepository(Entity1).find();

assertEquals(r1.status, 'fulfilled');
assertEquals(r2.status, 'rejected');

assertEquals(
	entities.length,
	1,
	'entity from transaction 1 should be saved',
);
assertEquals(
	entities[0].field,
	'a1',
	'should be entity from transaction 1',
);
```
```ts
@Entity()
class Entity1 {
	@PrimaryGeneratedColumn()
	id!: number;

	@Column({ type: 'varchar' })
	field: string;

	constructor(field: string) {
		this.field = field;
	}
}
```
```ts
async function transaction<T>(
	dataSource: DataSource,
	work: (em: EntityManager) => Promise<T>,
): Promise<T> {
	const queryRunner = dataSource.createQueryRunner();
	await queryRunner.startTransaction();

	try {
		const result = await work(queryRunner.manager);
		await queryRunner.commitTransaction();

		return result;
	} catch (error) {
		await queryRunner.rollbackTransaction();
		throw error;
	} finally {
		await queryRunner.release();
	}
}
```

I tried to create single file test that could be run directly via Deno 2, but that is unfortunately not possible because of the decorators which require a TS flag in a config file & `sqlite3`'s install script.

Reproduction repository: https://github.com/brunnerh/typeorm-isolation-issue

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    | Win 10   |
| Node.js version     | 18.19.0  |
| Typescript version  | 5.7.2  |
| TypeORM version     | 0.3.20  |

### Additional Context

Last working version: `0.3.18-dev.3cda7ec`
Broken since version: `0.3.18-dev.0f11739`

[Comparing the commits](https://github.com/typeorm/typeorm/compare/3cda7ec...0f11739) points to this PR as the source:
- #10210

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [ ] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [X] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by jorenvandeweyer** [comment]
I tested this on v0.3.20. I could not replicate it. There is exactly one entity being created as should be.

```typescript
const roleRepository = dataSource.getRepository(Role)

const { promise: wait, resolve: endWait } = Promise.withResolvers<void>()

const t1 = transaction(dataSource, async (em) => {
  const role1 = roleRepository.create({ name: randomBytes(8).toString('hex') })

  await em.save(role1)

  endWait()
  await t2.catch(() => {})
})

const t2 = transaction(dataSource, async (em) => {
  const role2 = roleRepository.create({ name: randomBytes(8).toString('hex') })

  await em.save(role2)
  await wait

  throw new Error('rollback transaction 2')
})

const [r1, r2] = await Promise.allSettled([t1, t2])
const entities = await roleRepository.find()

console.log(entities)

console.log(r1.status, r2.status)
```
```typescript
async function transaction<T> (
  dataSource: DataSource,
  work: (em: EntityManager) => Promise<T>
): Promise<T> {
  const queryRunner = dataSource.createQueryRunner()

  await queryRunner.startTransaction()

  try {
    const result = await work(queryRunner.manager)

    await queryRunner.commitTransaction()

    return result
  } catch (error) {
    await queryRunner.rollbackTransaction()

    throw error
  } finally {
    await queryRunner.release()
  }
}
```
output:
```bash
[
  Role {
    uuid: '65300444-fd17-45b5-8356-3ed9a8e41908',
    createdAt: 2024-12-30T09:11:11.375Z,
    updatedAt: 2024-12-30T09:11:11.375Z,
    name: 'c340c8d3bfdeb1b3',
    permissions: []
  }
]
fulfilled rejected

**Comment by brunnerh** [comment]
The issue appears to have been fixed by now, probably by #11264, which reverted the PR I identified as the probable cause of the issue.
